### ISSUE TITLE
eclim-debug-attach fails with "invalid project: "

### ISSUE THREAD

**Issue by ulises** [senny/emacs-eclim#232]
As the title says. 

I kick off an external gradle debug test, and while the JVM is paused listening for connections I run `M-x eclim-debug-attach` on a project java file buffer.

This fails with a message in the minibuffer saying: "invalid project: " and not much else.

**Comment by ulises** [comment]
I'll try to find the time sometime this weekend to trace this and come up with a fix.

**Comment by ulises** [comment]
Forgot to add:
* eclim 2.5.0
* Eclipse Mars 4.5.0
* Emacs 24.5 (running on OS X, installed via `brew`)

**Comment by kleewho** [comment]
I can confirm that I saw similar behaviour sometimes... couldn't find any pattern or anything actually that could help with fixing.

**Comment by kleewho** [comment]
By the way... I forgot about that old issue https://github.com/senny/emacs-eclim/issues/80

**Comment by ulises** [comment]
As seen in gitter:
````
    eclim-project-mode sets eclim-project-name (buffer-local var) to the project name
    if you find-file a file, then eclim-project-name is nil
    result: sadness
    solution: (add-hook 'java-mode-hook
                        (lambda () (setq eclim-project-name (eclim-project-name))))
    I haven't checked, but I suspect that if you open a file via eclim-project-mode, then this doesn't happen
````

**Comment by kleewho** [comment]
@ulises looking at the code there are two places that sets eclim-project-name. One is:
```el
(defun eclim--project-set-current ()
  (ignore-errors
    (setq eclim-project-name (eclim--project-current-line))))
```

which is not that interesting, but the other seems very promising:

```el
(defun eclim-project-name (&optional filename)
  "Returns this file's project name. If the optional argument
FILENAME is given, return that file's  project name instead."
  (labels ((get-project-name (file)
                             (eclim/execute-command "project_by_resource" ("-f" file))))
    (if filename
        (get-project-name filename)
      (or eclim-project-name
          (and buffer-file-name (setq eclim-project-name (get-project-name buffer-file-name)))))))
```

I think it was my mistake to use everywhere `eclim-project-name` as variable. I think in most (if not all) places we should use `(eclim-project-name)` instead.  

No need to use `add-hook`.

**Comment by kleewho** [comment]
Check this [PR](https://github.com/senny/emacs-eclim/pull/235). It should fix all issues with `eclim-project-name`

**Comment by ulises** [comment]
Apologies for the late comment. Patch LGTM, I wonder if it'd be clearer to do without the variable altogether as well?

**Comment by kleewho** [comment]
You mean without: `eclim--project-name`? There's an issue about poor performance of company completion #217  and it was identified (but when I wrote it I realized that I need to rethink this I) that some function that depends on `eclim-project-name` was the fault. So having this variable save us some time needed to call `eclim`. I hope fix for #217 is already in PR #237

**Comment by kleewho** [comment]
But... addressing your concerns I think we should add docstring for both variable and function that will directly say that variable is usually not what you want to use.
